2-advanced-topics/data-pipelines/intro_data_pipelines.ipynb

```diff
@@ -0,0 +1,1037 @@
+    "\n",
+    "In addition to this, we maintain a clear **lineage** of data -- which makes transparent the changes made to the raw data and what transformations have been applied to obtain the final dataset. In databricks it is possible to configure controls such that the lineage of each column in a table can be traced schematically. \n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Do we already use data pipelines? Or are they completely new?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the code snippet below, we obtain the some suibnational population data by making a requests call to a specific url from census.gov. Then, we manipulate this data using pandas to get it in the form we require and retaining the relevant columns. Finally, we *save* this data. This code is taken from [here](subnational_population.ipynb)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "database_name = \"data_pipelines_tutorial\"\n",
+    "URL = 'https://www2.census.gov/programs-surveys/international-programs/tables/time-series/prh/kenya.xlsx'\n",
+    "\n",
+    "df_raw = pd.read_excel(URL, sheet_name='2000 - 2040', skiprows=2, header=None)\n",
+    "# Read correct header\n",
+    "header = df_raw.iloc[1]\n",
+    "df_raw.columns = header\n",
+    "df_raw = df_raw.drop([0,1,2])\n",
+    "\n",
+    "# Extract Total population columns \n",
+    "df_pop_wide = df_raw[df_raw.ADM_LEVEL==1][['CNTRY_NAME', 'ADM1_NAME']+[x for x in header if 'BTOTL' in x]]\n",
+    "df_pop = pd.melt(df_pop_wide, id_vars=['CNTRY_NAME', 'ADM1_NAME'], var_name='year', value_name='population')\n",
+    "df_pop['year'] = df_pop['year'].str.extract(r'(\\d+)').astype(int)\n",
+    "df_pop.columns = ['country_name', 'adm1_name', 'year', 'population']\n",
+    "\n",
+    "# Modifications to the admin1 and county name and add data_source\n",
+    "name_standardization = {\n",
+    "    ''\n",
+    "}\n",
+    "df_pop['country_name'] = df_pop['country_name'].str.title()\n",
+    "df_pop['adm1_name'] = df_pop['adm1_name'].str.replace(r'[-/]+', ' ', regex=True).str.title()\n",
+    "df_pop = df_pop.astype({'year': 'int', 'population': 'int'})\n",
+    "df_pop = df_pop.sort_values(['adm1_name', 'year'], ignore_index=True)\n",
```

Comment:
```suggestion
    "df_pop = df_pop.sort_values(['adm1_name', 'year'], ignore_index=True).reset_index(drop=True)\n",
```
adding this suggestion so that printing the table below shows consecutive numbers in the index, otherwise this is going to confuse participants